// @irys/sdk/build/cjs/node/tokens/solana.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const arbundles_1 = require("arbundles");
const bignumber_js_1 = tslib_1.__importDefault(require("bignumber.js"));
const bs58_1 = tslib_1.__importDefault(require("bs58"));
const tweetnacl_1 = tslib_1.__importDefault(require("tweetnacl"));
const token_1 = require("../token");
const async_retry_1 = tslib_1.__importDefault(require("async-retry"));
const web3_js_1 = require("@solana/web3.js");
class SolanaConfig extends token_1.BaseNodeToken {
    constructor(config) {
        var _a, _b;
        super(config);
        this.minConfirm = 1;
        this.finality = "finalized";
        this.base = ["lamports", 1e9];
        this.finality = (_b = (_a = this === null || this === void 0 ? void 0 : this.opts) === null || _a === void 0 ? void 0 : _a.finality) !== null && _b !== void 0 ? _b : "finalized";
    }
    getProvider() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (!this.providerInstance) {
                this.providerInstance = new web3_js_1.Connection(this.providerUrl, {
                    confirmTransactionInitialTimeout: 60000,
                    commitment: this.finality,
                });
            }
            return this.providerInstance;
        });
    }
    getKeyPair() {
        let key = this.wallet;
        if (typeof key !== "string") {
            key = bs58_1.default.encode(Buffer.from(key));
        }
        return web3_js_1.Keypair.fromSecretKey(bs58_1.default.decode(key));
    }
    getTx(txId) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const connection = yield this.getProvider();
            const stx = yield connection.getTransaction(txId, { commitment: this.finality, maxSupportedTransactionVersion: 0 });
            if (!stx)
                throw new Error("Confirmed tx not found");
            const currentSlot = yield connection.getSlot(this.finality);
            if (!stx.meta)
                throw new Error(`Unable to resolve transaction ${txId}`);
            const amount = new bignumber_js_1.default(stx.meta.postBalances[1]).minus(new bignumber_js_1.default(stx.meta.preBalances[1]));
            const staticAccountKeys = stx.transaction.message.getAccountKeys().staticAccountKeys;
            const tx = {
                from: staticAccountKeys[0].toBase58(),
                to: staticAccountKeys[1].toBase58(),
                amount: amount,
                blockHeight: new bignumber_js_1.default(stx.slot),
                pending: false,
                confirmed: currentSlot - stx.slot >= 1,
            };
            return tx;
        });
    }
    ownerToAddress(owner) {
        return bs58_1.default.encode(owner);
    }
    sign(data) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return yield (yield this.getSigner()).sign(data);
        });
    }
    getSigner() {
        const keyp = this.getKeyPair();
        const keypb = bs58_1.default.encode(Buffer.concat([Buffer.from(keyp.secretKey), keyp.publicKey.toBuffer()]));
        return new arbundles_1.SolanaSigner(keypb);
    }
    verify(pub, data, signature) {
        return arbundles_1.SolanaSigner.verify(pub, data, signature);
    }
    getCurrentHeight() {
        var _a;
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return new bignumber_js_1.default((_a = (yield (yield this.getProvider()).getEpochInfo()).blockHeight) !== null && _a !== void 0 ? _a : 0);
        });
    }
    getFee(_amount, _to) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            // const connection = await this.getProvider()
            // const block = await connection.getRecentBlockhash();
            // const feeCalc = await connection.getFeeCalculatorForBlockhash(
            //     block.blockhash,
            // );
            // return new BigNumber(feeCalc.value.lamportsPerSignature);
            return new bignumber_js_1.default(5000); // hardcode it for now
        });
    }
    sendTx(data) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const connection = yield this.getProvider();
            try {
                return yield (0, web3_js_1.sendAndConfirmTransaction)(connection, data, [this.getKeyPair()], { commitment: this.finality });
            }
            catch (e) {
                if (e.message.includes("30.")) {
                    const txId = e.message.match(/[A-Za-z0-9]{87,88}/g);
                    if (!txId)
                        throw e;
                    try {
                        const conf = yield connection.confirmTransaction(txId[0], this.finality);
                        if (conf)
                            return undefined;
                        throw {
                            message: e.message,
                            txId: txId[0],
                        };
                    }
                    catch (e) {
                        throw {
                            message: e.message,
                            txId: txId[0],
                        };
                    }
                }
                throw e;
            }
        });
    }
    createTx(amount, to, _fee) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            // TODO: figure out how to manually set fees
            const keys = this.getKeyPair();
            const blockHashInfo = yield (0, async_retry_1.default)((bail) => tslib_1.__awaiter(this, void 0, void 0, function* () {
                var _a;
                try {
                    return yield (yield this.getProvider()).getLatestBlockhash();
                }
                catch (e) {
                    if ((_a = e.message) === null || _a === void 0 ? void 0 : _a.includes("blockhash"))
                        throw e;
                    else
                        bail(e);
                    throw new Error("Unreachable");
                }
            }), { retries: 3, minTimeout: 1000 });
            const transaction = new web3_js_1.Transaction(Object.assign(Object.assign({}, blockHashInfo), { feePayer: keys.publicKey }));
            transaction.add(web3_js_1.SystemProgram.transfer({
                fromPubkey: keys.publicKey,
                toPubkey: new web3_js_1.PublicKey(to),
                lamports: +new bignumber_js_1.default(amount).toNumber(),
            }));
            const transactionBuffer = transaction.serializeMessage();
            const signature = tweetnacl_1.default.sign.detached(transactionBuffer, keys.secretKey);
            transaction.addSignature(keys.publicKey, Buffer.from(signature));
            return { tx: transaction, txId: undefined };
        });
    }
    getPublicKey() {
        const key = this.getKeyPair();
        return key.publicKey.toBuffer();
    }
}
exports.default = SolanaConfig;
