// @ardrive/turbo-sdk/lib/esm/common/currency.js
export class ZeroDecimalCurrency {
    constructor(amount, type) {
        this.amount = amount;
        this.type = type;
    }
}
export class TwoDecimalCurrency {
    constructor(a, type) {
        this.a = a;
        this.type = type;
    }
    get amount() {
        return this.a * 100;
    }
}
// Two decimal currencies that are supported by the Turbo API
export const USD = (usd) => new TwoDecimalCurrency(usd, 'usd');
export const EUR = (eur) => new TwoDecimalCurrency(eur, 'eur');
export const GBP = (gbp) => new TwoDecimalCurrency(gbp, 'gbp');
export const CAD = (cad) => new TwoDecimalCurrency(cad, 'cad');
export const AUD = (aud) => new TwoDecimalCurrency(aud, 'aud');
export const INR = (inr) => new TwoDecimalCurrency(inr, 'inr');
export const SGD = (sgd) => new TwoDecimalCurrency(sgd, 'sgd');
export const HKD = (hkd) => new TwoDecimalCurrency(hkd, 'hkd');
export const BRL = (brl) => new TwoDecimalCurrency(brl, 'brl');
// Zero decimal currencies that are supported by the Turbo API
export const JPY = (jpy) => new ZeroDecimalCurrency(jpy, 'jpy');
