// @ardrive/turbo-sdk/lib/cjs/common/token/ethereum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthereumToken = exports.ETHToTokenAmount = exports.weiToTokenAmount = void 0;
/**
 * Copyright (C) 2022-2023 Permanent Data Solutions, Inc. All Rights Reserved.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
const bignumber_js_1 = require("bignumber.js");
const ethers_1 = require("ethers");
const logger_js_1 = require("../logger.js");
const weiToTokenAmount = (wei) => wei;
exports.weiToTokenAmount = weiToTokenAmount;
const ETHToTokenAmount = (eth) => new bignumber_js_1.BigNumber(eth).times(1e18).valueOf();
exports.ETHToTokenAmount = ETHToTokenAmount;
class EthereumToken {
    constructor({ logger = new logger_js_1.TurboWinstonLogger(), gatewayUrl = 'https://cloudflare-eth.com/', pollingOptions = {
        maxAttempts: 10,
        pollingIntervalMs: 4000,
        initialBackoffMs: 10000,
    }, } = {}) {
        this.logger = logger;
        this.gatewayUrl = gatewayUrl;
        this.pollingOptions = pollingOptions;
        this.rpcProvider = new ethers_1.ethers.JsonRpcProvider(gatewayUrl);
    }
    async createAndSubmitTx({ target, tokenAmount, signer, }) {
        // convert wei to eth
        const eth = tokenAmount.shiftedBy(-18);
        const txId = await signer.sendTransaction({
            target,
            amount: eth,
            provider: this.rpcProvider,
        });
        return {
            id: txId,
            target,
        };
    }
    async pollForTxBeingAvailable({ txId, }) {
        await new Promise((resolve) => setTimeout(resolve, this.pollingOptions.initialBackoffMs));
        let attempts = 0;
        while (attempts < this.pollingOptions.maxAttempts) {
            try {
                const tx = await this.rpcProvider.getTransaction(txId);
                if (tx) {
                    return;
                }
            }
            catch (e) {
                this.logger.debug('Error polling for tx', { txId, e });
            }
            await new Promise((resolve) => setTimeout(resolve, this.pollingOptions.pollingIntervalMs));
            attempts++;
        }
        throw new Error('Transaction not found after polling!');
    }
}
exports.EthereumToken = EthereumToken;
