// @ardrive/turbo-sdk/lib/cjs/common/factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TurboBaseFactory = void 0;
/**
 * Copyright (C) 2022-2023 Permanent Data Solutions, Inc. All Rights Reserved.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
const arbundles_1 = require("arbundles");
const signer_js_1 = require("../node/signer.js");
const types_js_1 = require("../types.js");
const signer_js_2 = require("../web/signer.js");
const logger_js_1 = require("./logger.js");
const payment_js_1 = require("./payment.js");
const index_js_1 = require("./token/index.js");
const turbo_js_1 = require("./turbo.js");
const upload_js_1 = require("./upload.js");
class TurboBaseFactory {
    static setLogLevel(level) {
        this.logger.setLogLevel(level);
    }
    static setLogFormat(format) {
        this.logger.setLogFormat(format);
    }
    static unauthenticated({ paymentServiceConfig = {}, uploadServiceConfig = {}, } = {}) {
        const paymentService = new payment_js_1.TurboUnauthenticatedPaymentService({
            ...paymentServiceConfig,
            logger: this.logger,
        });
        const uploadService = new upload_js_1.TurboUnauthenticatedUploadService({
            ...uploadServiceConfig,
            logger: this.logger,
        });
        return new turbo_js_1.TurboUnauthenticatedClient({
            uploadService,
            paymentService,
        });
    }
    static getSigner(providedSigner, providedPrivateKey, token) {
        let signer;
        if (providedSigner !== undefined) {
            signer = providedSigner;
        }
        else if (providedPrivateKey !== undefined) {
            if (token === 'solana') {
                signer = new arbundles_1.HexSolanaSigner(providedPrivateKey);
            }
            else if (token === 'ethereum') {
                if (!(0, types_js_1.isEthPrivateKey)(providedPrivateKey)) {
                    throw new Error('An Ethereum private key must be provided for EthereumSigner.');
                }
                signer = new arbundles_1.EthereumSigner(providedPrivateKey);
            }
            else {
                if (!(0, types_js_1.isJWK)(providedPrivateKey)) {
                    throw new Error('A JWK must be provided for ArweaveSigner.');
                }
                signer = new arbundles_1.ArweaveSigner(providedPrivateKey);
            }
        }
        else {
            throw new Error('A privateKey or signer must be provided.');
        }
        if (typeof window !== 'undefined') {
            return new signer_js_2.TurboWebArweaveSigner({
                signer,
                logger: this.logger,
            });
        }
        return new signer_js_1.TurboNodeSigner({
            signer,
            logger: this.logger,
        });
    }
    static authenticated({ privateKey, signer: providedSigner, paymentServiceConfig = {}, uploadServiceConfig = {}, token, tokenMap, gatewayUrl, tokenTools, }) {
        if (!token) {
            if (providedSigner) {
                // Derive token from signer if not provided
                if (providedSigner instanceof arbundles_1.EthereumSigner) {
                    token = 'ethereum';
                }
                else if (providedSigner instanceof arbundles_1.HexSolanaSigner) {
                    token = 'solana';
                }
                else {
                    token = 'arweave';
                }
            }
            else {
                token = 'arweave';
            }
        }
        const turboSigner = this.getSigner(providedSigner, privateKey, token);
        if (!tokenTools) {
            if (tokenMap && token === 'arweave') {
                tokenTools = tokenMap.arweave;
            }
            tokenTools = index_js_1.defaultTokenMap[token]?.({
                gatewayUrl,
                logger: this.logger,
            });
        }
        const paymentService = new payment_js_1.TurboAuthenticatedPaymentService({
            ...paymentServiceConfig,
            signer: turboSigner,
            logger: this.logger,
            token,
            tokenTools,
        });
        const uploadService = new upload_js_1.TurboAuthenticatedUploadService({
            ...uploadServiceConfig,
            signer: turboSigner,
            logger: this.logger,
            token,
        });
        return new turbo_js_1.TurboAuthenticatedClient({
            uploadService,
            paymentService,
        });
    }
}
exports.TurboBaseFactory = TurboBaseFactory;
TurboBaseFactory.logger = new logger_js_1.TurboWinstonLogger();
