// @ardrive/turbo-sdk/lib/cjs/common/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TurboWinstonLogger = void 0;
/**
 * Copyright (C) 2022-2023 Permanent Data Solutions, Inc. All Rights Reserved.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
const winston_1 = require("winston");
const version_js_1 = require("../version.js");
class TurboWinstonLogger {
    constructor({ level = 'info', logFormat = 'simple', } = {}) {
        this.logger = (0, winston_1.createLogger)({
            level,
            defaultMeta: { client: 'turbo-sdk', version: version_js_1.version },
            silent: level === 'none',
            format: getLogFormat(logFormat),
            transports: [new winston_1.transports.Console()],
        });
    }
    info(message, ...args) {
        this.logger.info(message, ...args);
    }
    warn(message, ...args) {
        this.logger.warn(message, ...args);
    }
    error(message, ...args) {
        this.logger.error(message, ...args);
    }
    debug(message, ...args) {
        this.logger.debug(message, ...args);
    }
    setLogLevel(level) {
        this.logger.level = level;
    }
    setLogFormat(logFormat) {
        this.logger.format = getLogFormat(logFormat);
    }
}
exports.TurboWinstonLogger = TurboWinstonLogger;
function getLogFormat(logFormat) {
    return winston_1.format.combine((0, winston_1.format)((info) => {
        if (info.stack && info.level !== 'error') {
            delete info.stack;
        }
        return info;
    })(), winston_1.format.errors({ stack: true }), // Ensure errors show a stack trace
    winston_1.format.timestamp(), logFormat === 'json' ? winston_1.format.json() : winston_1.format.simple());
}
