// @ardrive/turbo-sdk/lib/cjs/common/upload.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TurboAuthenticatedUploadService = exports.TurboUnauthenticatedUploadService = exports.defaultUploadServiceURL = exports.developmentUploadServiceURL = void 0;
const http_js_1 = require("./http.js");
const logger_js_1 = require("./logger.js");
exports.developmentUploadServiceURL = 'https://upload.ardrive.dev';
exports.defaultUploadServiceURL = 'https://upload.ardrive.io';
class TurboUnauthenticatedUploadService {
    constructor({ url = exports.defaultUploadServiceURL, retryConfig, logger = new logger_js_1.TurboWinstonLogger(), token = 'arweave', }) {
        this.token = token;
        this.logger = logger;
        this.httpService = new http_js_1.TurboHTTPService({
            url: `${url}/v1`,
            retryConfig,
            logger: this.logger,
        });
    }
    async uploadSignedDataItem({ dataItemStreamFactory, dataItemSizeFactory, signal, }) {
        const fileSize = dataItemSizeFactory();
        this.logger.debug('Uploading signed data item...');
        // TODO: add p-limit constraint or replace with separate upload class
        return this.httpService.post({
            endpoint: `/tx/${this.token}`,
            signal,
            data: dataItemStreamFactory(),
            headers: {
                'content-type': 'application/octet-stream',
                'content-length': `${fileSize}`,
            },
        });
    }
}
exports.TurboUnauthenticatedUploadService = TurboUnauthenticatedUploadService;
// NOTE: to avoid redundancy, we use inheritance here - but generally prefer composition over inheritance
class TurboAuthenticatedUploadService extends TurboUnauthenticatedUploadService {
    constructor({ url = exports.defaultUploadServiceURL, retryConfig, signer, logger, token, }) {
        super({ url, retryConfig, logger, token });
        this.signer = signer;
    }
    async uploadFile({ fileStreamFactory, fileSizeFactory, signal, dataItemOpts, }) {
        const { dataItemStreamFactory, dataItemSizeFactory } = await this.signer.signDataItem({
            fileStreamFactory,
            fileSizeFactory,
            dataItemOpts,
        });
        const signedDataItem = dataItemStreamFactory();
        const fileSize = dataItemSizeFactory();
        this.logger.debug('Uploading signed data item...');
        // TODO: add p-limit constraint or replace with separate upload class
        return this.httpService.post({
            endpoint: `/tx/${this.token}`,
            signal,
            data: signedDataItem,
            headers: {
                'content-type': 'application/octet-stream',
                'content-length': `${fileSize}`,
            },
        });
    }
}
exports.TurboAuthenticatedUploadService = TurboAuthenticatedUploadService;
